libbpf-tools/gethostlatency.c

```diff
@@ -0,0 +1,286 @@
+	if (func_off < 0) {
+		warn("could not find getaddrinfo in %s\n", libc_path);
+		return -1;
+	}
+	obj->links.hanlde_entry =
+		bpf_program__attach_uprobe(obj->progs.hanlde_entry, false,
+					   traced_pid ?: -1, libc_path, func_off);
+	err = libbpf_get_error(obj->links.hanlde_entry);
+	if (err) {
+		warn("failed to attach getaddrinfo: %d\n", err);
+		return -1;
+	}
+	obj->links.hanlde_return =
+		bpf_program__attach_uprobe(obj->progs.hanlde_return, true,
+					   traced_pid ?: -1, libc_path, func_off);
+	err = libbpf_get_error(obj->links.hanlde_return);
+	if (err) {
+		warn("failed to attach getaddrinfo: %d\n", err);
+		return -1;
+	}
+
+	func_off = get_elf_func_offset(libc_path, "gethostbyname");
+	if (func_off < 0) {
+		warn("Could not find gethostbyname in %s\n", libc_path);
+		return -1;
+	}
+	obj->links.hanlde_entry =
+		bpf_program__attach_uprobe(obj->progs.hanlde_entry, false,
+					   traced_pid ?: -1, libc_path, func_off);
+	err = libbpf_get_error(obj->links.hanlde_entry);
+	if (err) {
+		warn("failed to attach gethostbyname: %d\n", err);
+		return -1;
+	}
+	obj->links.hanlde_return =
+		bpf_program__attach_uprobe(obj->progs.hanlde_return, true,
+					   traced_pid ?: -1, libc_path, func_off);
+	err = libbpf_get_error(obj->links.hanlde_return);
+	if (err) {
+		warn("failed to attach gethostbyname: %d\n", err);
+		return -1;
+	}
+
+	func_off = get_elf_func_offset(libc_path, "gethostbyname2");
+	if (func_off < 0) {
+		warn("Could not find gethostbyname2 in %s\n", libc_path);
+		return -1;
+	}
+	obj->links.hanlde_entry =
+		bpf_program__attach_uprobe(obj->progs.hanlde_entry, false,
+					   traced_pid ?: -1, libc_path, func_off);
+	err = libbpf_get_error(obj->links.hanlde_entry);
+	if (err) {
+		warn("failed to attach gethostbyname2: %d\n", err);
+		return -1;
+	}
+	obj->links.hanlde_return =
+		bpf_program__attach_uprobe(obj->progs.hanlde_return, true,
+					   traced_pid ?: -1, libc_path, func_off);
+	err = libbpf_get_error(obj->links.hanlde_return);
+	if (err) {
+		warn("failed to attach gethostbyname2: %d\n", err);
+		return -1;
+	}
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	struct gethostlatency_bpf *obj;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = gethostlatency_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open BPF object\n");
+		return 1;
+	}
+
+	obj->rodata->targ_tgid = traced_pid;
+
+	err = gethostlatency_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	err = attach_uprobes(obj);
+	if (err) {
+		warn("failed to attach BPF programs: %s\n", strerror(-err));
```

Comment:
`attach_uprobes()` seems to return -1 for errors, so I don't think calling strerror() is right here. Could probably just drop the error message from this warning -- attach_uprobes() already prints out a helpful message